src/ts_catalog/dimension_partition.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+#include "cache.h"
+#include <postgres.h>
+#include <access/heapam.h>
+#include <access/xact.h>
+#include <catalog/catalog.h>
+#include <commands/tablecmds.h>
+#include <lib/rbtree.h>
+#include <nodes/parsenodes.h>
+#include <utils/array.h>
+#include <utils/palloc.h>
+#include <utils/rel.h>
+#include <utils/builtins.h>
+#include <utils/snapmgr.h>
+
+#include "ts_catalog/catalog.h"
+#include "dimension.h"
+#include "dimension_slice.h"
+#include "dimension_partition.h"
+#include "hypertable_cache.h"
+#include "scanner.h"
+
+#include "compat/compat.h"
+
+static ScanIterator
+ts_dimension_partition_scan_iterator_create(LOCKMODE lockmode)
+{
+	ScanIterator it = ts_scan_iterator_create(DIMENSION_PARTITION, lockmode, CurrentMemoryContext);
+	it.ctx.flags |= SCANNER_F_NOEND_AND_NOCLOSE;
+
+	return it;
+}
+
+static void
+ts_dimension_partition_scan_iterator_set_dimension_id(ScanIterator *it, int32 dimension_id,
+													  const ScanTupLock *tuplock)
+{
+	it->ctx.index = catalog_get_index(ts_catalog_get(),
+									  DIMENSION_PARTITION,
+									  DIMENSION_PARTITION_DIMENSION_ID_RANGE_START_IDX);
+	ts_scan_iterator_scan_key_reset(it);
+	ts_scan_iterator_scan_key_init(
+		it,
+		Anum_dimension_partition_dimension_id_range_start_idx_dimension_id,
+		BTEqualStrategyNumber,
+		F_INT4EQ,
+		Int32GetDatum(dimension_id));
+	it->ctx.tuplock = tuplock;
+}
+
+/*
+ * Find a partition in the tree.
+ *
+ * The tree is built based on the start of each partition's range.
+ *
+ * Used also to lookup the matching partition for a particular point, which
+ * will be disguised as a partition with range_start==range_end.
+ *
+ * The "key" is the user-passed argument that we are matching against the
+ * tree.
+ *
+ * The "node" is the node in the tree that we are comparing against.
+ */
+static int
+dimpart_cmp(const RBTNode *key, const RBTNode *node, void *arg)
```

Comment:
Sure, array and binary search should work the same, just different implementation. I can look quickly on whether it would simplify the code or not but otherwise probably not worthwhile making the change since it won't make a huge difference.

Reply:
Switched to array+binary search. Made the code a bit cleaner, so thanks for the suggestion.